chain/client/src/stateless_validation/chunk_endorsement/tracker_v2.rs

```diff
@@ -0,0 +1,132 @@
+            let mut chunk_endorsements = self.chunk_endorsements.lock().unwrap();
+            let entry = chunk_endorsements.get_or_insert_mut(key, || HashMap::new());
+            entry.insert(account_id.clone(), endorsement);
+        };
+        Ok(())
+    }
+
+    /// Called by block producer.
+    /// Returns ChunkEndorsementsState::Endorsed if node has enough signed stake for the chunk
+    /// represented by chunk_header.
+    /// Signatures have the same order as ordered_chunk_validators, thus ready to be included in a block as is.
+    /// Returns ChunkEndorsementsState::NotEnoughStake if chunk doesn't have enough stake.
+    pub(crate) fn compute_chunk_endorsements(
+        &self,
+        chunk_header: &ShardChunkHeader,
+    ) -> Result<ChunkEndorsementsState, Error> {
+        let shard_id = chunk_header.shard_id();
+        let epoch_id =
+            self.epoch_manager.get_epoch_id_from_prev_block(chunk_header.prev_block_hash())?;
+        let protocol_version = self.epoch_manager.get_epoch_protocol_version(&epoch_id)?;
+        if !ProtocolFeature::StatelessValidation.enabled(protocol_version) {
+            // Return an empty array of chunk endorsements for older protocol versions.
+            return Ok(ChunkEndorsementsState::Endorsed(None, vec![]));
+        }
+
+        let height_created = chunk_header.height_created();
+        let key = ChunkProductionKey { shard_id, epoch_id, height_created };
+
+        let chunk_validator_assignments = self.epoch_manager.get_chunk_validator_assignments(
+            &epoch_id,
+            chunk_header.shard_id(),
+            chunk_header.height_created(),
+        )?;
+
+        // Get the chunk_endorsements for the chunk from our cache.
+        // Note that these chunk endorsements are already validated as part of process_chunk_endorsement.
+        // We can safely rely on the following details
+        //    1. The chunk endorsements are from valid chunk_validator for this chunk.
+        //    2. The chunk endorsements signatures are valid.
+        //    3. We still need to validate if the chunk_hash matches the chunk_header.chunk_hash()
+        let mut chunk_endorsements = self.chunk_endorsements.lock().unwrap();
+        let Some(entry) = chunk_endorsements.get_mut(&key) else {
+            // Early return if no chunk_endorsements found in our cache.
+            return Ok(ChunkEndorsementsState::NotEnoughStake(None));
+        };
+        entry.retain(|_, endorsement| endorsement.chunk_hash() == &chunk_header.chunk_hash());
+
+        let endorsement_stats =
+            chunk_validator_assignments.compute_endorsement_stats(&entry.keys().collect());
+
+        // Check whether the current set of chunk_validators have enough stake to include chunk in block.
+        if !endorsement_stats.has_enough_stake() {
+            return Ok(ChunkEndorsementsState::NotEnoughStake(Some(endorsement_stats)));
```

Comment:
for my knowledge, when there is not enough stake, will this function be called later with more endorsements? maybe indicate this in the function comments?